core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaUpdateRecipesTranslator.java

```diff
@@ -323,4 +342,41 @@ private static class GroupedItem {
         ItemDefinition id;
         int count;
     }
+
+    private List<RecipeData> getUpgradeRecipes(GeyserSession session) {
+        ArrayList<RecipeData> recipes = new ArrayList<>();
+        ItemMapping template = session.getItemMappings().getStoredItems().upgradeTemplate();
+
+        for (String identifier : NETHERITE_TRANSFORM) {
+            var recipe = org.cloudburstmc.protocol.bedrock.data.inventory.crafting.recipe.SmithingTransformRecipeData.of(identifier + "_smithing",
+                    getDescriptorFromId(session, template.getBedrockIdentifier()),
+                    getDescriptorFromId(session, identifier.replace("netherite", "diamond")),
+                    getDescriptorFromId(session, "minecraft:netherite_ingot"),
+                    getItemData(session, identifier),
+                    "smithing_table",
+                    session.getLastRecipeNetId().getAndIncrement());
+            recipes.add(recipe);
+        }
+        return recipes;
+    }
+
+    private ItemDescriptorWithCount getDescriptorFromId(GeyserSession session, String bedrockId) {
+        for (ItemDefinition itemDefinition : session.getItemMappings().getItemDefinitions().values()) {
+            if (itemDefinition.getIdentifier().equals(bedrockId)) {
+                return ItemDescriptorWithCount.fromItem(ItemData.builder().definition(itemDefinition).count(1).build());
+            }
+        }
+        GeyserImpl.getInstance().getLogger().debug("Unable to find item with identifier " + bedrockId);
+        return ItemDescriptorWithCount.EMPTY;
+    }
+
+    private ItemData getItemData(GeyserSession session, String bedrockID) {
```

Comment:
If these two methods you introduced don't exist in a similar fashion somewhere else, it might make sense to move them somewhere else since they aren't really recipe-specific

Isn't there already a method or registry somewhere that converts a bedrock string ID to BedrockIdentifier?

Reply:
They don't exist in that form; i would move them. Had to create a similar one here: https://github.com/GeyserMC/Geyser/commit/b1f04a9012e9c64c9931a86f7045aa4cebb769c4. Personally, i find the whole ItemMapping v ItemDefinition v ItemData etc fairly confusing, so i might have missed a method that does this.